File: desktop-mend/src/test/java/commands/MergeTest.java
package commands;

import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.desktop.commands.Merge;
import org.apache.commons.lang3.tuple.Pair;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.runners.MockitoJUnitRunner;
import org.mockito.stubbing.Answer;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class MergeTest extends TestBase {
    private Merge merge;

    @Captor
    private ArgumentCaptor<Pair<File, File>> logFilesCaptor;

    @Before
    public void setup() {
        super.setup();
        merge = new Merge(strings, log, fileResolveHelper, mergeHelper, osDao);
    }

    @Test
    public void mergeToFirst() throws IOException, CorruptSettingsException {
        mergeInPlaceTest(true);
    }

    @Test
    public void mergeToSecond() throws IOException, CorruptSettingsException {
        mergeInPlaceTest(false);
    }

    @Test
    public void mergeToNewFile() throws IOException, CorruptSettingsException {
        String logA = "logA";
        String logB = "logB";
        String logC = "logC";
        when(osDao.fileExists(any())).thenReturn(true);
        when(osDao.fileIsFile(any())).thenReturn(true);
        resolveAnyFile();
        merge.execute(Arrays.asList(logA, logB, logC));
        ArgumentCaptor<File> fileCaptor = ArgumentCaptor.forClass(File.class);
        verify(mergeHelper).mergeLogFilesToNew(logFilesCaptor.capture(), fileCaptor.capture());

        Assert.assertEquals(logA, logFilesCaptor.getValue().getLeft().getName());
        Assert.assertEquals(logB, logFilesCaptor.getValue().getRight().getName());
        Assert.assertEquals(logC, fileCaptor.getValue().getName());
    }

    @Test
    public void tooManyArgs() {
        wrongArgNumTest(4);
    }

    @Test
    public void tooFewArgs() {
        wrongArgNumTest(2);
    }

    public void wrongArgNumTest(int argNum) {
        merge.execute(IntStream.of(argNum).mapToObj(i -> "arg").collect(Collectors.toList()));
        verify(err).println(strings.getf("General.invalidArgNum", Merge.COMMAND_NAME));
    }

    private void mergeInPlaceTest(boolean first) throws IOException, CorruptSettingsException {
        String logA = "logA";
        String logB = "logB";
        when(osDao.fileExists(any())).thenReturn(true);
        when(osDao.fileIsFile(any())).thenReturn(true);
        resolveAnyFile();
        merge.execute(Arrays.asList(first ? Merge.FIRST_FLAG : Merge.SECOND_FLAG, logA, logB));
        verify(mergeHelper).mergeToFirstOrSecond(logFilesCaptor.capture(), eq(first));

        Assert.assertEquals(logA, logFilesCaptor.getValue().getLeft().getName());
        Assert.assertEquals(logB, logFilesCaptor.getValue().getRight().getName());
    }

    private void resolveAnyFile() throws IOException, CorruptSettingsException {
        when(fileResolveHelper.resolveAsLogFilePath(anyString())).thenAnswer(
                (Answer<File>) invocation
                        -> new File((String)invocation.getArguments()[0]));
    }
}

File: desktop-mend/src/test/java/commands/MergeTest.kt
package commands

import co.samco.mend4.desktop.commands.Merge
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.eq
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.io.File
import java.util.stream.Collectors
import java.util.stream.IntStream

class MergeTest : TestBase() {
    private var merge: Merge? = null

    @Before
    override fun setup() {
        super.setup()
        merge = Merge(
            strings = strings,
            log = log,
            fileResolveHelper = fileResolveHelper,
            settingsHelper = settingsHelper,
            mergeHelper = mergeHelper
        )
    }

    @Test
    fun testCommandWithSettingsDependencies() {
        super.testCommandWithNoSettingsDependencies(merge!!)
    }

    @Test
    fun mergeToFirst() {
        mergeInPlaceTest(true)
    }

    @Test
    fun mergeToSecond() {
        mergeInPlaceTest(false)
    }

    @Test
    fun mergeToNewFile() {
        val logA = "logA"
        val logB = "logB"
        val logC = "logC"
        whenever(osDao.exists(any())).thenReturn(true)
        whenever(osDao.exists(any())).thenReturn(true)
        whenever(fileResolveHelper.ensureLogNameHasFileExtension(eq(logC))).thenReturn(logC)
        resolveAnyFile()
        merge!!.execute(listOf(logA, logB, logC))
        val fileCaptor = argumentCaptor<File>()
        val logFilesCaptor = argumentCaptor<Pair<File, File>>()
        verify(mergeHelper).mergeLogFilesToNew(logFilesCaptor.capture(), fileCaptor.capture())
        Assert.assertEquals(logA, logFilesCaptor.firstValue.first.name)
        Assert.assertEquals(logB, logFilesCaptor.firstValue.second.name)
        Assert.assertEquals(logC, fileCaptor.firstValue.name)
    }

    @Test
    fun tooManyArgs() {
        wrongArgNumTest(4)
    }

    @Test
    fun tooFewArgs() {
        wrongArgNumTest(2)
    }

    private fun wrongArgNumTest(argNum: Int) {
        merge!!.execute(IntStream.of(argNum).mapToObj { i: Int -> "arg" }
            .collect(Collectors.toList()))
        verify(err).println(strings.getf("General.invalidArgNum", Merge.COMMAND_NAME))
    }

    private fun mergeInPlaceTest(first: Boolean) {
        val logA = "logA"
        val logB = "logB"
        whenever(osDao.exists(any())).thenReturn(true)
        whenever(osDao.exists(any())).thenReturn(true)
        resolveAnyFile()
        merge!!.execute(
            listOf(
                if (first) Merge.FIRST_FLAG else Merge.SECOND_FLAG,
                logA,
                logB
            )
        )
        val logFilesCaptor = argumentCaptor<Pair<File, File>>()
        verify(mergeHelper).mergeToFirstOrSecond(logFilesCaptor.capture(), eq(first))
        Assert.assertEquals(logA, logFilesCaptor.firstValue.first.name)
        Assert.assertEquals(logB, logFilesCaptor.firstValue.second.name)
    }

    private fun resolveAnyFile() {
        whenever(fileResolveHelper.resolveAsLogFilePath(any())).thenAnswer { inv ->
            File(inv.arguments[0] as String)
        }
    }
}